myhpom/tasks.py

```diff
@@ -0,0 +1,119 @@
+import json
+import requests
+from django.core.mail import send_mail
+from django.conf import settings
+from django.template.loader import get_template
+from django.template import Context
+from celery import shared_task
+from celery.task import Task
+from celery.signals import task_failure
+from myhpom.models import AdvanceDirective, DocumentUrl, CloudFactoryRun, CloudFactoryUnit
+
+
+@shared_task
+class CloudFactorySubmitAdvanceDirectiveRun(Task):
+    def run(self, ad_id, line_id, document_host, callback_url):
+        """
+        ## Arguments
+        * ad_id = the database id for the AdvanceDirective
+        * line_id = the id of the CloudFactory production line to submit this request to.
+        * document_host = the host (incl. protocol/scheme) to use for the DocumentURL
+        * callback_url = the URL to use with the callback request
+
+        ## Task Process
+        - create an expiring DocumentUrl for the AdvanceDirective
+        - submit the DocumentUrl to the CloudFactory (test) endpoint
+        - store the status of the submitted document
+        - if the AdvanceDirective no longer exists, abort the task without an Exception.
+            - the AdvanceDirective might no longer exist by the time celery picks it up,
+            - or the document it points to might have been removed.
+        - if an error occurs in this process, send support email.
+        """
+        cf_run = CloudFactoryRun.objects.create(line_id=line_id, callback_url=callback_url or "")
+
+        # If the AdvanceDirective has been deleted, abort the task with a message
+        # -- no need to admin support email, this is an expected case.
+        try:
+            ad = AdvanceDirective.objects.get(id=ad_id)
+        except:
+            cf_run.status = 'ABORTED'
+            cf_run.message = 'AdvanceDirective(id=%r) no longer exists.' % ad_id
+            cf_run.save()
+            return cf_run.data
+
+        # Any exception in the rest of the task should result in support email
+        document_url = DocumentUrl.objects.create(advancedirective=ad)
+        CloudFactoryUnit.objects.create(
+            run=cf_run, input=self.create_unit_input(document_url, document_host)
+        )
+        response = self.post_run(cf_run.post_data)
+        response_data = response.json()
```

Comment:
Is this in case CloudFactory is offline or throws a 500? Maybe figure out which response.status_codes are likely to be non-json. For the status_codes that give us json, it should throw an error and email us if the response data isn't json – because it means CloudFactory has radically changed their API and we need to talk.